kong/clustering/rpc/socket.lua

```diff
@@ -0,0 +1,287 @@
+        local cb = self.manager.callbacks.callbacks[payload.method]
+        if not cb then
+          local res, err = self.outgoing:push({
+            jsonrpc = "2.0",
+            id = payload.id,
+            ["error"] = {
+              code = jsonrpc.METHOD_NOT_FOUND,
+              message = "Method not found",
+            }
+          })
+          if not res then
+            return nil, "unable to handle ping: " .. err
+          end
+
+          goto continue
+        end
+
+        -- call dispatch
+
+        assert(ngx.timer.at(0, function(premature)
+          if premature then
+            return
+          end
+
+          local res, err = cb(self.node_id, unpack(payload.params))
+          if not res then
+            ngx_log(ngx_WARN, "[rpc] RPC callback failed: ", err)
+
+            res, err = self.outgoing:push({
+              jsonrpc = "2.0",
+              id = payload.id,
+              ["error"] = {
+                code = jsonrpc.SERVER_ERROR,
+                message = tostring(err),
+              }
+            })
+            if not res then
+              ngx_log(ngx_WARN, "[rpc] unable to push RPC call error: ", err)
+            end
+
+            return
+          end
+
+          -- success
+          res, err = self.outgoing:push({
+            jsonrpc = "2.0",
+            id = payload.id,
+            result = res,
+          })
+          if not res then
+            ngx_log(ngx_WARN, "[rpc] unable to push RPC call result: ", err)
+          end
+        end))
+
+      else
+        -- response
+        local cb = self.interest[payload.id]
+        self.interest[payload.id] = nil -- edge trigger only once
+
+        if not cb then
+          ngx_log(ngx_WARN, "[rpc] no interest for RPC response id: ", payload.id, ", dropping it")
+
+          goto continue
+        end
+
+        local res, err = cb(payload)
+        if not res then
+          ngx_log(ngx_WARN, "[rpc] RPC response interest handler failed: id: ",
+                  payload.id, ", err: ", err)
+        end
+      end
+
+      ::continue::
+    end
+  end)
+
+  self.write_thread = ngx.thread.spawn(function()
+    while not exiting() do
+      local payload, err = self.outgoing:pop(5)
+      if err then
+        return nil, err
+      end
+
+      if payload then
+        if payload == PING_TYPE then
+          local _, err = self.wb:send_ping()
+          if err then
+            return nil, "failed to send PING frame to peer: " .. err
+
+          else
+            ngx_log(ngx_DEBUG, "[rpc] sent PING frame to peer")
+          end
+
+        elseif payload == PONG_TYPE then
+          local _, err = self.wb:send_pong()
+          if err then
+            return nil, "failed to send PONG frame to peer: " .. err
+
+          else
+            ngx_log(ngx_DEBUG, "[rpc] sent PONG frame to peer")
+          end
+
+        else
+          assert(type(payload) == "table")
+
+          local bytes, err = self.wb:send_binary(cjson_encode(payload))
```

Comment:
should we check the return value of `cjson.encode`?

Reply:
Added `assert`. For compression, we will add Snappy support for the protocol, not Gzip and not necessarily a first stage goal.